Reset Stash next_free to 0 and clear data when size becomes 0

spearman opened the pull request:
So that a newly allocated stash behaves the same as a recently
cleared stash (either using clear or the last element has been
removed with take).

Did a quick trial with a type that implements Drop to make sure nothing is being dropped twice, looked okay.

spearman committed:
set Stash next_free to 0 when len becomes 0

So that a newly allocated stash behaves the same as a recently
cleared stash (either using clear or the last element has been
removed with take).

spearman committed:
also clear data, added reset_zero test

Stebalien commented:
Is there any particular reason you want this? I intentionally make no guarantees about ordering.

spearman commented:
I'm maintaining two stashes in parallel on different threads, using assigned handles for create/update/destroy messages (single producer, single consumer, messages handled in order); sometimes the producer will drop (clearing the consumer) and be re-initialized at a later time.

If this change breaks the no-guarantees assumption I could probably just send a special 'clear' message to the consumer and re-assign the empty stash to `= Stash::new()` to keep them in synch.

Stebalien commented:
I don't even guarantee that that will work; key choice may be non-deterministic and may change at any time. Hence the description:

> A fast map for when one doesn't care about choosing the keys.

However, given a compelling use case, I could consider changing this. Are you trying to build a pipeline?

spearman commented:
More like a kind of CRUD interface without the 'R' (i.e. the data flow is one-way). The endpoint is an OpenGL 3.x rendering thread so the commands are used to acquire and release rendering resources.

Actually it would make more sense to just make the receiving end a VecMap, so it's not really a compelling use case, no. There might be some kind of use case with distributed data where messages are bidirectional.

Stebalien commented on src/stash/mod.rs:
```diff
@@ -399,6 +407,8 @@ impl<V> Stash<V> {
             self.next_free = i;
```
Could you document the fact that a cleared stash operates identically to a new stash?

Stebalien commented on src/stash/mod.rs:
```diff
@@ -351,20 +351,28 @@ impl<V> Stash<V> {
 
     /// Take an item from a slot (if non empty).
     pub fn take(&mut self, index: usize) -> Option<V> {
+        let mut out = None;
```
Do you really need *these* changes? IMO, it's enough to make *cleared* stashes act like new stashes (not stashes that happen to become empty).

Stebalien reviewed (changes requested):
So, I've decided to merge the clear behavior but hold off on the take behavior. It makes sense to "reset" a stash on clear but I don't see any value in resetting it if it happens to become empty. Does this make sense?

spearman committed:
only set next_free to 0 after clear() but not when len becomes 0

spearman committed:
document clear() behavior

spearman commented:
Reverted changes to `take()`, rewrote the test function to only test for cleared stashes behaving like new stashes and added a note to the `clear()` function documentation.

Just FYI I ended up not even using this fork as I realized what I was doing was stupid :).

Stebalien commented on src/stash/mod.rs:
```diff
@@ -356,6 +356,7 @@ impl<V> Stash<V> {
                 Entry::Empty(free_slot) => {
                     // Just put it back.
                     *entry = Entry::Empty(free_slot);
+                    return None;
```
No point in leaving this change here.

Stebalien reviewed (changes requested):
Please squash (this really only needs one commit) and avoid making unnecessary changes.

Stebalien commented:
@spearman 
> Just FYI I ended up not even using this fork as I realized what I was doing was stupid :).

No problem. If you're short on time, I can just fix this up myself (I don't want to drag you through a review for something you don't use).

spearman commented:
squashed to #4
